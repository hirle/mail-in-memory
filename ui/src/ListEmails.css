.error {
    font-size: 36px;
    color: red;
}

thead {
    background-color: #333;
    color: #fff;
}

td {
    border: 1px solid #333;
}