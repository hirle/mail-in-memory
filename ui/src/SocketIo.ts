import io from 'socket.io-client';

export type NotifyEvent = (data: any) => void;

export default class SocketIo {

    private socket?: SocketIOClient.Socket;
    private observers: Map<string, NotifyEvent[]>;

    constructor() {
        this.socket = undefined;
        this.observers = new Map<string, NotifyEvent[]>();
    }

    public startOn(path: string) {
        this.socket = io(path);
    }

    public stop() {
        // TODO code the stop
        console.log('not implemented yet');
    }

    public addObserver( event: string, notifyEvent: NotifyEvent ){
        if( ! this.observers.has(event)) {
            this.observers.set(event, []);
            this.socket?.on(event, this.notify.bind(this, event))
        }
        this.observers.get(event)!.push(notifyEvent);
    }

    public isConnected(): boolean {
        return !!this.socket && this.socket.connected;
    }

    public notify(event: string, data: any){
        if( this.observers.has(event) ) {
            this.observers.get(event)!.forEach( notifyEvent => notifyEvent(data));
        }
    }

}