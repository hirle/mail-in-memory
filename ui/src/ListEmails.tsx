import React from 'react';
import {Mail} from '@mail-in-memory/model';
import {SocketMessages} from '@mail-in-memory/model';
import {GetLastEmails} from './api';
import SocketIo from './SocketIo';
import './ListEmails.css';

interface ComponentProps {
    socketIo: SocketIo
}
  
enum ComponentStatus {
    Loading,
    Ready,
    Error
}
  
  interface ComponentState {
      connected: boolean,
      status: ComponentStatus,
      mails: Mail[]
  }

  export default class CurrentTemperature
    extends React.Component<ComponentProps, ComponentState> {

    constructor(props: ComponentProps) {
        super(props);
        this.state = {status: ComponentStatus.Loading, connected: false, mails: []};
    }
      
    componentDidMount() {
        GetLastEmails()
          .then( mails => {        
            this.setState({
              status: ComponentStatus.Ready,
              mails,
              connected:this.props.socketIo.isConnected()
            });
  
            this.setUpSocketIO();
          })
          .catch( err => {
            console.log(err);
            this.setState({status: ComponentStatus.Error, mails: []});
          });
    }

    setUpSocketIO() {
        this.props.socketIo.addObserver(SocketMessages.Connect, this.onConnect.bind(this));
        this.props.socketIo.addObserver(SocketMessages.NewMail, this.onNewMail.bind(this));
        this.props.socketIo.addObserver(SocketMessages.Disconnect, this.onDisconnect.bind(this));
    }
    
    onNewMail( ) {
        GetLastEmails()
          .then( mails => {        
            this.setState({
              status: ComponentStatus.Ready,
              mails,
              connected:this.props.socketIo.isConnected()
            });
        })
        .catch( err => {
            console.log(err);
            this.setState({status: ComponentStatus.Error, mails: []});
        });
    
    }
    
    onConnect() {
    this.setState({connected: true});
    }

    onDisconnect() {
    this.setState({connected: false});
    }

    render() {
        switch(this.state.status)  {
          case ComponentStatus.Loading: return <div>Loading...</div>;
          case ComponentStatus.Ready: return (
              <table>
                <thead>
                    <tr>
                        <th>From</th>
                        <th>To</th>
                        <th>Subject</th>
                        <th>Body</th>
                        <th>TimeStamp</th>
                    </tr>
                </thead>
                <tbody>
                  { this.state.mails.map( (mail, index) => (<tr key={index}>
                      <td >{mail.fromAddress}</td>
                      <td >{mail.toAddress}</td>
                      <td >{mail.subject||''}</td>
                      <td >{mail.body}</td>
                      <td >{mail.mailTimestamp.toLocaleString()}</td>
                      </tr>)
                ) }
                </tbody>
              </table>);
          default: return <div className="error">Can't connect</div>;
        }
    }
}
    