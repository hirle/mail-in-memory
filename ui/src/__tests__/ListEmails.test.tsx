import renderer from "react-test-renderer";
import SocketIo from "../SocketIo";
import ListEmails from '../ListEmails';

describe('ListEmails', () => {
    it('should show loading', () => {
        const mySocketIo = new SocketIo();
        const component = renderer.create(<ListEmails socketIo={mySocketIo} />);
        let mustBeLoading = component.toJSON();
        expect(mustBeLoading).toMatchSnapshot();
    });
});