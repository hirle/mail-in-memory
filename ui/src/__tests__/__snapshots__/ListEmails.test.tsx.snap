// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListEmails should show loading 1`] = `
<div>
  Loading...
</div>
`;
